Update README.md (#356)
mattermost/mattermost-plugin-github: README.md

A GitHub plugin for Mattermost. Supports GitHub SaaS and Enterprise versions.

## Table of Contents

 - [1. Audience](#1-audience)
 - [2. License](#2-license)
 - [3. About the GitHub Plugin](#3-about-the-github-plugin)
 - [4. Before You Start](#4-before-you-start)
 - [5. Configuration](#6-configuration)
 - [6. Using the Plugin](#7-using-the-plugin)
 - [7. Onboarding Your Users](#8-onboarding-your-users)
 - [8. Slash Commands](#9-slash-commands)
 - [9. Configuration Notes in High Availability Mode](#4-configuration-notes-in-high-availability-mode)
 - [10. Frequently Asked Questions](#6-frequently-asked-questions)
 - [11. Development](#5-development)

![GitHub plugin screenshot](https://user-images.githubusercontent.com/13119842/54380268-6adab180-4661-11e9-8470-a9c615c00041.png)

## Audience

This guide is intended for Mattermost System Admins setting up the GitHub plugin and Mattermost users who want information about the plugin functionality. For more information about contributing to this plugin, visit the [Development section](https://github.com/mattermost/mattermost-plugin-github#development).

## License

> This repository is licensed under the Apache 2.0 License.

Reviewer: Nit pick: Maybe link to `LICENSE`?

Author: Added a link to the Apache 2.0 license wording.

3. (Optional) **Enable Private Repositories:** Allow the plugin to receive notifications from private repositories by setting this value to `true`.
4. (**Enterprise Only**) **Enterprise Base URL** and **Enterprise Upload URL**: Set these values to your GitHub Enterprise URLs, e.g. `https://github.example.com`. The Base and Upload URLs are often the same. When enabled, existing users must reconnect their accounts to gain access to private repositories. Affected users will be notified by the plugin once private repositories are enabled.
5. Hit **Save**.
6. Go to **System Console > Plugins > Management** and click **Enable** to enable the GitHub plugin.

You're all set!

## Using the Plugin

Once configuration is complete, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command and its output isn't visible to other members of the channel.

## Onboarding Your Users

When you’ve tested the plugin and confirmed it’s working, notify your team so they can connect their GitHub account to Mattermost and get started. Copy and paste the text below, edit it to suit your requirements, and send it out.

> Hi team, 

> > We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command is only visible to you.. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.

Reviewer: What does `The command is only visible to you` mean?

Author: When you type the command in a channel, other members of the channel don't see it afaik. I was a bit worried about that when I first set it up. We do also reference that in other documentation (which I'll need to find to cross-check).

Reviewer: Should we talk about the command-response then? The term `command` usually refers to the slash command i.e. the user input.

Author: Ah yes, that works much better.

> > We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command is only visible to you.. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.

Reviewer: What does `The command is only visible to you` mean?

A GitHub plugin for Mattermost. Supports GitHub SaaS and Enterprise versions.

## Table of Contents

 - [1. Audience](#1-audience)
 - [2. License](#2-license)
 - [3. About the GitHub Plugin](#3-about-the-github-plugin)
 - [4. Before You Start](#4-before-you-start)
 - [5. Check your Version](#5-check-your-version)
 - [6. Configuration](#6-configuration)
 - [7. Using the Plugin](#7-using-the-plugin)
 - [8. Onboarding Your Users](#8-onboarding-your-users)
 - [9. Slash Commands](#9-slash-commands)
 - [10. Configuration Notes in High Availability Mode](#4-configuration-notes-in-high-availability-mode)
 - [11. Development](#5-development)

>  - [12. Frequently Asked Questions](#6-frequently-asked-questions)

Author: It should be - however, that is an assumption based on the doc I copied this from. I still need to verify it.

>  - [12. Frequently Asked Questions](#6-frequently-asked-questions)

Author: Oh, I misread. No, it's not auto-generated, I added the sections manually. Apologies!

> ![GitHub plugin screenshot](https://user-images.githubusercontent.com/13119842/54380268-6adab180-4661-11e9-8470-a9c615c00041.png)

Author: I was going to mention in this PR that it'd be great to update it. I'm happy to give it a bash today.

> ![GitHub plugin screenshot](https://user-images.githubusercontent.com/13119842/54380268-6adab180-4661-11e9-8470-a9c615c00041.png)

Author: I haven't yet done this screenshot update - @hanzei, do you perhaps have one that we can use in the interim?

> ![GitHub plugin screenshot](https://user-images.githubusercontent.com/13119842/54380268-6adab180-4661-11e9-8470-a9c615c00041.png)

Reviewer: @justinegeffen What do you think about using 
![oie_ho3algV3kVzV](https://user-images.githubusercontent.com/16541325/93596960-893e0d80-f9ba-11ea-8ec5-d7610c98864c.png)
?

> ![GitHub plugin screenshot](https://user-images.githubusercontent.com/13119842/54380268-6adab180-4661-11e9-8470-a9c615c00041.png)

Author: I'm very happy with that. Thank you!

## Audience

> This guide is intended for Mattermost System Admins setting up the GitHub plugin and Mattermost users who want information about the plugin functionality.

Author: Yes, good idea.

> This guide is intended for Mattermost System Admins setting up the GitHub plugin and Mattermost users who want information about the plugin functionality.

Author: ```suggestion
This guide is intended for Mattermost System Admins setting up the GitHub plugin and Mattermost users who want information about the plugin functionality. For more information about contributing to this plugin, visit the [Development section](https://github.com/mattermost/mattermost-plugin-github#development).
```

> ## License

Author: Need to confirm that this is correct license information for this repo.

> This repository is licensed under the Apache 2.0 License, except for the [server/enterprise](server/enterprise) directory which is licensed under the [Mattermost Source Available License](LICENSE.enterprise). See [Mattermost Source Available License](https://docs.mattermost.com/overview/faq.html#mattermost-source-available-license) to learn more.

Author: ```suggestion
This repository is licensed under the Apache 2.0 License.
```

After your System Admin has [configured the GitHub plugin](#Configuration), run `/github connect` in a Mattermost channel to connect your Mattermost and GitHub accounts.

Once connected, you'll have access to the following features:

* __Daily reminders__ - The first time you log in to Mattermost each day, get a post letting you know what issues and pull requests need your attention.
* __Notifications__ - Get a direct message in Mattermost when someone mentions you, requests your review, comments on or modifies one of your pull requests/issues, or assigns you on GitHub.
* __Post actions__ - Create a GitHub issue from a post or attach a post message to an issue. Hover over a post to reveal the post actions menu and click **More Actions (...)**.
* __Sidebar buttons__ - Stay up-to-date with how many reviews, unread messages, assignments, and open pull requests you have with buttons in the Mattermost sidebar
* __Slash commands__ - Interact with the GitHub plugin using the `/github` slash command

## Before You Start

> This guide assumes that you have a GitHub account and that you're a Mattermost System Admin.

Author: ```suggestion
This guide assumes that you have a GitHub account, that you're a Mattermost System Admin, and you're running Mattermost v5.12 or higher.
```

> If you’re running Mattermost v5.11 or earlier, first download the latest release of this plugin from the [releases page of this GitHub repository](https://github.com/mattermost/mattermost-plugin-github/releases) and upload it to your Mattermost instance [following this documentation](https://docs.mattermost.com/administration/plugins.html#plugin-uploads).

Author: @hanzei, is 5.11 unsupported? If so, we can either remove this section or we can have a version with 5.11 and a current version?

## Check your version

> In Mattermost 5.12 and later, the GitHub plugin is pre-packaged and no steps are required for installation. You can go directly to [Configuration](#configuration).

Author: ```suggestion
In Mattermost 5.12 and later, the GitHub plugin is bundled and no steps are required for installation. You can go directly to [Configuration](#configuration).
```

In Mattermost 5.11 and earlier, follow these steps:

> 1. Go to https://github.com/mattermost/mattermost-plugin-github/releases to download the latest release file in zip or tar.gz format.

Author: Hmm, not sure what happened there because you are absolutely correct.

2. Upload the file through **System Console > Plugins > Management**, or manually upload it to the Mattermost server under plugin directory. See [documentation](https://docs.mattermost.com/administration/plugins.html#set-up-guide) for more details.
3. Once complete, follow the steps provided in the Configuration section.

## Configuration

Configuration is started in GitHub and completed in Mattermost. 

> **Note:** If you're using GitHub Enterprise, replace all GitHub links below with your GitHub Enterprise URL.

Author: @hanzei, what are your thoughts around having an EE version of this as well in GitBook?

Author: Fair point, I'm on board with that.

6. Hit **Save**.

### Step 2: Create a Webhook in GitHub

You must create a webhook for each organization you want to receive notifications for or subscribe to.

1. In **System Console > Plugins > GitHub**, generate a new value for **Webhook Secret**. Copy it, as you will use it in a later step.
2. Hit **Save** to save the secret.
3. Go to the **Settings** page of your GitHub organization you want to send notifications from, then select **Webhooks** in the sidebar.
4. Click **Add Webhook**.
5. Set the following values:
   - **Payload URL:** `https://your-mattermost-url.com/plugins/github/webhook`, replacing `https://your-mattermost-url.com` with your Mattermost URL.
   - **Content Type:** `application/json`
   - **Secret:** the webhook secret you copied previously.
6. Select **Let me select individual events** for "Which events would you like to trigger this webhook?".
7. Select the following events: `Branch or Tag creation`, `Branch or Tag deletion`, `Issue comments`, `Issues`, `Pull requests`, `Pull request review`, `Pull request review comments`, `Pushes`.
7. Hit **Add Webhook** to save it.

> Repeat this process if you have multiple organizations.

Author: Updated.

3. (Optional) **Enable Private Repositories:** Allow the plugin to receive notifications from private repositories by setting this value to `true`.
4. (**Enterprise Only**) **Enterprise Base URL** and **Enterprise Upload URL**: Set these values to your GitHub Enterprise URLs, e.g. `https://github.example.com`. The Base and Upload URLs are often the same. When enabled, existing users must reconnect their accounts to gain access to private repositories. Affected users will be notified by the plugin once private repositories are enabled.
5. Hit **Save**.
6. Go to **System Console > Plugins > Management** and click **Enable** to enable the GitHub plugin.

You're all set!

## Using the Plugin

Once configuration is complete, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command is only visible to you.

## Onboarding Your Users

When you’ve tested the plugin and confirmed it’s working, notify your team so they can connect their GitHub account to Mattermost and get started. Copy and paste the text below, edit it to suit your requirements, and send it out.

**Hi team, 

> We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, follow these steps to connect your GitHub account to your Mattermost client. Then, take a look at the slash commands section for details about how to use the plugin.**

Author: ```suggestion
We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command is only visible to you.. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.**
```
To

## Slash Commands

* __Subscribe to a respository__ - Use `/github subscribe` to subscribe a Mattermost channel to receive notifications for new pull requests, issues, branch creation, and more in a GitHub repository.

   - For instance, to post notifications for issues, issue comments, and pull requests matching the label `Help Wanted` from `mattermost/mattermost-server`, use:
   ```
   /github subscribe mattermost/mattermost-server issues,pulls,issue_comments,label:"Help Wanted"
   ```
  - The following flags are supported:
   - `--exclude-org-member`: events triggered by organization members will not be delivered. It will be locked to the organization provided in the plugin configuration and it will only work for users whose membership is public. Note that organization members and collaborators are not the same.
   
* __Get to do items__ - Use `/github todo` to get an ephemeral message with items to do in GitHub, including a list of unread messages and pull requests awaiting your review.
* __Update settings__ - Use `/github settings` to update your settings for notifications and daily reminders.

> * __And more!__ - Run `/github help` to see what else the slash command can do.

Reviewer: What do you think about this idea?

> * __And more!__ - Run `/github help` to see what else the slash command can do.

Author: It's a great idea - I did add a slash command point to this; not sure why it's not displaying though. I'll check my commits.

> * __And more!__ - Run `/github help` to see what else the slash command can do.

Author: Looks like I didn't add it as I'd thought. On it now.

### Configuration Notes in High Availability Mode

> If you are running Mattermost v5.11 or earlier in [High Availability mode](https://docs.mattermost.com/deployment/cluster.html), please review the following:

Author: Should we drop just this line, or the whole section? The way it's structured it reads like "if you're using 5.11 or earlier... then the rest of this section applies to you". Does this content apply to versions 5.12 and up?

> If you are running Mattermost v5.11 or earlier in [High Availability mode](https://docs.mattermost.com/deployment/cluster.html), please review the following:

Reviewer: No, it doesn't apply to 5.12 and up. Hence, I think we should drop the whole section.

> If you are running Mattermost v5.11 or earlier in [High Availability mode](https://docs.mattermost.com/deployment/cluster.html), please review the following:

Reviewer: No, it doesn't apply to 5.12 and up. Hence, I think we should drop the whole section.

> If you are running Mattermost v5.11 or earlier in [High Availability mode](https://docs.mattermost.com/deployment/cluster.html), please review the following:

Author: Removed.

A GitHub plugin for Mattermost. Supports GitHub SaaS and Enterprise versions.

## Table of Contents

>  - [1. Audience](#1-audience)

Reviewer: These don't seem to work when testing on https://github.com/mattermost/mattermost-plugin-github/blob/justinegeffen-github-revision/README.md. Do they work on you local instance?

 - [License](#license)
 - [About the GitHub Plugin](#about-the-github-plugin)
 - [Before You Start](#before-you-start)
 - [Configuration](#configuration)
 - [Using the Plugin](#using-the-plugin)
 - [Onboarding Your Users](#onboarding-your-users)
 - [Slash Commands](#slash-commands)
 - [Frequently Asked Questions](#frequently-asked-questions)
 - [Development](#development)

> ![GitHub plugin screenshot](assets/github_mattermost.png)

Reviewer: Files in`assets` as bundled with the plugin release which is not needed. Could you please put the file into `images` similar to how we do it in https://github.com/mattermost/mattermost-plugin-jira?

Author: I don't see an ``images`` directory in this repo - wondering if the Jira repo has one because of it being on Gitbook?

Reviewer: Yes, you need to create an `image` directory. The assets for GitBook are stored in https://github.com/mattermost/mattermost-plugin-jira/tree/master/docs/.gitbook/assets

## Audience

This guide is intended for Mattermost System Admins setting up the GitHub plugin and Mattermost users who want information about the plugin functionality. For more information about contributing to this plugin, visit the [Development section](https://github.com/mattermost/mattermost-plugin-github#development).

## License

> This repository is licensed under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0).

Reviewer: Actually, I was suggesting to link to https://github.com/mattermost/mattermost-plugin-github/blob/master/LICENSE. 0/5 of what the better link target is.

## About the GitHub Plugin

> The Mattermost GitHub plugin uses a webhook to connect your GitHub account to Mattermost, to listen for incoming GitHub events. Events notifications are via DM in Mattermost. The Events don’t need separate configuration and include: 

Reviewer 2: ```suggestion
The Mattermost GitHub plugin uses a webhook to connect your GitHub account to Mattermost to listen for incoming GitHub events. Events notifications are via DM in Mattermost. The Events don’t need separate configuration and include: 
```

1. In **System Console > Plugins > GitHub**, generate a new value for **Webhook Secret**. Copy it, as you will use it in a later step.
2. Hit **Save** to save the secret.
3. Go to the **Settings** page of your GitHub organization you want to send notifications from, then select **Webhooks** in the sidebar.
4. Click **Add Webhook**.
5. Set the following values:
   - **Payload URL:** `https://your-mattermost-url.com/plugins/github/webhook`, replacing `https://your-mattermost-url.com` with your Mattermost URL.
   - **Content Type:** `application/json`
   - **Secret:** the webhook secret you copied previously.
6. Select **Let me select individual events** for "Which events would you like to trigger this webhook?".
7. Select the following events: `Branch or Tag creation`, `Branch or Tag deletion`, `Issue comments`, `Issues`, `Pull requests`, `Pull request review`, `Pull request review comments`, `Pushes`.
7. Hit **Add Webhook** to save it.

If you have multiple organizations, repeat the process starting from step 3 to create a webhook for each organization.

### Step 3: Configure the Plugin in Mattermost

> If you have an existing Mattermost user account with the name `github` the plugin will post using the `github` account but without a `BOT` tag.

Reviewer 2: ```suggestion
If you have an existing Mattermost user account with the name `github`, the plugin will post using the `github` account but without a `BOT` tag.
```

3. (Optional) **Enable Private Repositories:** Allow the plugin to receive notifications from private repositories by setting this value to `true`.
4. (**Enterprise Only**) **Enterprise Base URL** and **Enterprise Upload URL**: Set these values to your GitHub Enterprise URLs, e.g. `https://github.example.com`. The Base and Upload URLs are often the same. When enabled, existing users must reconnect their accounts to gain access to private repositories. Affected users will be notified by the plugin once private repositories are enabled.
5. Hit **Save**.
6. Go to **System Console > Plugins > Management** and click **Enable** to enable the GitHub plugin.

You're all set!

## Using the Plugin

Once configuration is complete, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub.

## Onboarding Your Users

When you’ve tested the plugin and confirmed it’s working, notify your team so they can connect their GitHub account to Mattermost and get started. Copy and paste the text below, edit it to suit your requirements, and send it out.

> Hi team, 

> > We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.

Reviewer: That link does't work right now. Can you confirm that it will after the merge?

Author: I've adjusted it.

## Slash Commands

> * __Autocomplete slash commands__ - Explore all the available slash commands by typing `/` in the text input box - the autocomplete suggestions help by providing a format example in black text and a short description of the slash command in grey text. Visit the [executing commands](https://docs.mattermost.com/help/messaging/executing-commands.html) documentation for more details. You can also create custom slash commands. Visit the [slash commands](https://docs.mattermost.com/developer/slash-commands.html) documentation for more information and a list of default slash commands.

Reviewer: 1/5 Mentioning custom slash command seems a bit weird to me. The audience of this documents are admins who are setting up the GitHub plugin, not developers of slash commands.

* __Subscribe to a respository__ - Use `/github subscribe` to subscribe a Mattermost channel to receive notifications for new pull requests, issues, branch creation, and more in a GitHub repository.

   - For instance, to post notifications for issues, issue comments, and pull requests matching the label `Help Wanted` from `mattermost/mattermost-server`, use:
   ```
   /github subscribe mattermost/mattermost-server issues,pulls,issue_comments,label:"Help Wanted"
   ```
  - The following flags are supported:

>    - `--exclude-org-member`: events triggered by organization members will not be delivered. It will be locked to the organization provided in the plugin configuration and it will only work for users whose membership is public. Note that organization members and collaborators are not the same.

Reviewer 2: ```suggestion
     - `--exclude-org-member`: events triggered by organization members will not be delivered. It will be locked to the organization provided in the plugin configuration and it will only work for users whose membership is public. Note that organization members and collaborators are not the same.
```

account with the name `github`.

**Note:** For `v0.9.0` and earlier of the GitHub plugin, instead of using bot accounts, set the username the plugin is attached to in **System Console > Plugins > GitHub**.

#### Generate a Key
  
Open **System Console > Plugins > GitHub** and do the following:

1. Generate a new value for **At Rest Encryption Key**.
2. (Optional) **GitHub Organization:** Lock the plugin to a single GitHub organization by setting this field to the name of your GitHub organization.
3. (Optional) **Enable Private Repositories:** Allow the plugin to receive notifications from private repositories by setting this value to `true`.
4. (**Enterprise Only**) **Enterprise Base URL** and **Enterprise Upload URL**: Set these values to your GitHub Enterprise URLs, e.g. `https://github.example.com`. The Base and Upload URLs are often the same. When enabled, existing users must reconnect their accounts to gain access to private repositories. Affected users will be notified by the plugin once private repositories are enabled.
5. Hit **Save**.
6. Go to **System Console > Plugins > Management** and click **Enable** to enable the GitHub plugin.

You're all set!

## Using the Plugin

> Once configuration is complete, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command and its output isn't visible to other members of the channel.

Author: ```suggestion
Once configuration is complete, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub.
```

## Onboarding Your Users

When you’ve tested the plugin and confirmed it’s working, notify your team so they can connect their GitHub account to Mattermost and get started. Copy and paste the text below, edit it to suit your requirements, and send it out.

> Hi team, 

> > We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. The command is only visible to you.. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.

Author: ```suggestion
> We've set up the Mattermost GitHub plugin, so you can get notifications from GitHub in Mattermost. To get started, run the `/github connect` slash command from any channel within Mattermost to connect your Mattermost account with GitHub. Then, take a look at the [slash commands](https://github.com/mattermost/mattermost-plugin-github#slash-commands) section for details about how to use the plugin.
```

> 

Reviewer: For the dev section I would love to just link to the docs added in https://github.com/mattermost/mattermost-developer-documentation/pull/661. I've just merged that PR and I will take some hours until the changes are visible.

What do you think about a linking to the dev docs?

Author: Yes, awesome idea!! I'm 100% happy with that. Could we link there for all the sections that reference development in the READMEs (e.g., Jira, Zoom, etc)?

Author: Would this be the correct link: https://developers.mattermost.com/extend/plugins/developer-workflow/?

Reviewer: Yes, the section could be used for all plugins. We should just copy&past it to every README we update.

Reviewer: https://developers.mattermost.com/extend/plugins/developer-setup/ and https://developers.mattermost.com/extend/plugins/developer-workflow/ are the page that should be linked here.

Reviewer: Yes, the section could be used for all plugins. We should just copy&past it to every README we update.

Author: Totally agree - a single source is way better and easier to maintain. Thanks for doing those!

> 

Author: Good suggestion, I've moved it.
